File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
from collections import deque

N = int(input())

for i in range(N):
    I = int(input())

    q = deque([])

    start_row, start_col = map(int, input().split(' '))
    end_row, end_col = map(int, input().split(' '))

    if (start_row==end_row and start_col==end_col):
        print(0)
        continue

    q.append((start_row, start_col))

    maps = [[0]*I for _ in range(I)]
    cnt = 1

    while q:
        row, col = q.popleft()


        for nrow, ncol in [(row-1,col-2),(row-2,col-1),(row-2,col+1),(row-1,col+2),(row+1,col+2),(row+2,col+1),(row+2,col-1),(row+1,col-2)]:
        
            if (0<=nrow<I and 0<=ncol<I and maps[nrow][ncol] == 0): 
                maps[nrow][ncol] = maps[row][col] + 1
                q.append((nrow,ncol))
                if (nrow==end_row and ncol==end_col):
                    print(maps[end_row][end_col])
                    break
                

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.cc
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int main() {
    int N;
    cin >> N;

    vector<pair<int, int>> moves = {
        {-1, -2}, {-2, -1}, {-2, 1}, {-1, 2},
        {1, 2}, {2, 1}, {2, -1}, {1, -2}
    };

    for (int i = 0; i < N; i++) {
        int I;
        cin >> I;
        int start_row, start_col, end_row, end_col;
        cin >> start_row >> start_col >> end_row >> end_col;

        if (start_row == end_row && start_col == end_col) {
            cout << 0 << '\n';
            continue;
        }

        vector<vector<int>> maps(I, vector<int>(I, 0));
        queue<pair<int, int>> q;
        q.push({start_row, start_col});
        bool found = false;

        while (!q.empty() && !found) {
            int row = q.front().first;
            int col = q.front().second;
            q.pop();

            for (auto [dr, dc] : moves) {
                int nrow = row + dr;
                int ncol = col + dc;

                if (nrow >= 0 && nrow < I && ncol >= 0 && ncol < I && maps[nrow][ncol] == 0) {
                    maps[nrow][ncol] = maps[row][col] + 1;

                    if (nrow == end_row && ncol == end_col) {
                        cout << maps[nrow][ncol] << '\n';
                        found = true;
                        break;
                    }

                    q.push({nrow, ncol});
                }
            }
        }
    }

    return 0;
}
